explorer/interpreter/interpreter.cpp

```diff
@@ -1603,24 +1694,62 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
         }
       }
     case StatementKind::Break: {
-      CARBON_CHECK(act.pos() == 0);
+      StatementAction& statement_action = cast<StatementAction>(act);
+      if (act.pos() == 0) {
+        auto destructor_calls =
+            CollectVariablesToDestruct(stmt, CaptureVariables::Block);
+        statement_action.add_destructor_calls(destructor_calls);
+      }
+      if (statement_action.HasDestructorCalls()) {
+        auto call_item = statement_action.PopDestructorCall();
+        return CallDestructor(call_item.first, call_item.second);
+      }
```

Comment:
Rather than try to call the destructors before unwinding the stack, I think we should call the destructors as part of unwinding the stack. After all, that's a big part of what "unwinding the stack" is supposed to mean, and I think it will substantially simplify the code. It will probably make the code more robust as well, because the logic for running the destructors will be built directly into the `ActionStack` logic, instead of having to find everywhere in the interpreter that potentially ends a scope and add destruction logic there. For example, I'm not confident this PR correctly destroys temporary objects that are created by implicit conversions when calling a function (see [here](https://github.com/carbon-language/carbon-lang/blob/c096d010b96c5c526d7d35620699ee3364fbb2e2/explorer/interpreter/interpreter.cpp#L772)), because that temporary scope is associated with an expression, not a statement.

Here's roughly how I envision this working:
- Define a new kind of `Action` called `CleanupAction` which owns a `RuntimeScope`, and have each step of the `CleanupAction` destroy one element of the `RuntimeScope` (the `pos` member tells you which element to destroy).
- Inside `ActionStack`, every time we pop an `Action` off the stack that isn't a `CleanupAction`, if its `scope()` isn't empty, we push a `CleanupAction` onto the stack that cleans up the scope of the `Action` we just popped. If we pop more than one `Action` at a time, as in `PopScopes` and the `Unwind` methods, we have to push a `CleanupAction` for each `Action` that we popped, and in the right order.

Does that make sense?

Reply:
I had a similar idea, however I failed in several places.... However, now I have a bit more understanding of the code and could try again!

Reply:
Is implemented!